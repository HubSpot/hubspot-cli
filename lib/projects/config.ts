import fs from 'fs-extra';
import path from 'path';
import findup from 'findup-sync';
import { logger } from '@hubspot/local-dev-lib/logger';
import { getAbsoluteFilePath, getCwd } from '@hubspot/local-dev-lib/path';

import { ProjectConfig } from '../../types/Projects';
import { PROJECT_CONFIG_FILE } from '../constants';
import { lib } from '../../lang/en';
import { uiCommandReference } from '../ui';
import { EXIT_CODES } from '../enums/exitCodes';

export function writeProjectConfig(
  configPath: string,
  config: ProjectConfig
): boolean {
  try {
    fs.ensureFileSync(configPath);
    fs.writeFileSync(configPath, JSON.stringify(config, null, 2));
    logger.debug(`Wrote project config at ${configPath}`);
  } catch (e) {
    logger.debug(e);
    return false;
  }
  return true;
}

export function getIsInProject(dir?: string): boolean {
  const configPath = getProjectConfigPath(dir);
  return !!configPath;
}

function getProjectConfigPath(dir?: string): string | null {
  const projectDir = dir ? getAbsoluteFilePath(dir) : getCwd();

  const configPath = findup(PROJECT_CONFIG_FILE, {
    cwd: projectDir,
    nocase: true,
  });

  return configPath;
}

export interface LoadedProjectConfig {
  projectDir: string | null;
  projectConfig: ProjectConfig | null;
}

export async function getProjectConfig(
  dir?: string
): Promise<LoadedProjectConfig> {
  const configPath = getProjectConfigPath(dir);
  if (!configPath) {
    return { projectConfig: null, projectDir: null };
  }

  try {
    const config = fs.readFileSync(configPath);
    const projectConfig: ProjectConfig = JSON.parse(config.toString());
    return {
      projectDir: path.dirname(configPath),
      projectConfig,
    };
  } catch (e) {
    logger.error('Could not read from project config');
    return { projectConfig: null, projectDir: null };
  }
}

export function validateProjectConfig(
  projectConfig: ProjectConfig | null,
  projectDir: string | null
): asserts projectConfig is ProjectConfig {
  if (!projectConfig || !projectDir) {
    logger.error(
      lib.projects.validateProjectConfig.configNotFound(
        uiCommandReference('hs project create')
      )
    );
    return process.exit(EXIT_CODES.ERROR);
  }

  if (!projectConfig.name || !projectConfig.srcDir) {
    logger.error(lib.projects.validateProjectConfig.configMissingFields);
    return process.exit(EXIT_CODES.ERROR);
  }

  const resolvedPath = path.resolve(projectDir, projectConfig.srcDir);
  if (!resolvedPath.startsWith(projectDir)) {
    const projectConfigFile = path.relative(
      '.',
      path.join(projectDir, PROJECT_CONFIG_FILE)
    );
    logger.error(
      lib.projects.validateProjectConfig.srcOutsideProjectDir(
        projectConfigFile,
        projectConfig.srcDir
      )
    );
    return process.exit(EXIT_CODES.ERROR);
  }

  if (!fs.existsSync(resolvedPath)) {
    logger.error(
      lib.projects.validateProjectConfig.srcDirNotFound(
        projectConfig.srcDir,
        projectDir
      )
    );

    return process.exit(EXIT_CODES.ERROR);
  }
}
