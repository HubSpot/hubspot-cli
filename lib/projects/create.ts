import { logger } from '@hubspot/local-dev-lib/logger';
import { fetchRepoFile } from '@hubspot/local-dev-lib/api/github';
import { RepoPath } from '@hubspot/local-dev-lib/types/Github';
import {
  HUBSPOT_PROJECT_COMPONENTS_GITHUB_PATH,
  PROJECT_COMPONENT_TYPES,
} from '../constants';
import { EXIT_CODES } from '../enums/exitCodes';
import { i18n } from '../lang';
import {
  ProjectTemplate,
  ComponentTemplate,
  ProjectTemplateRepoConfig,
} from '../../types/Projects';
import { debugError } from '../errorHandlers/index';

const i18nKey = 'lib.projects.create';

export const EMPTY_PROJECT_TEMPLATE_NAME = 'no-template';
const PROJECT_TEMPLATE_PROPERTIES = ['name', 'label', 'path', 'insertPath'];

export async function getProjectComponentListFromRepo(
  githubRef: string
): Promise<ComponentTemplate[]> {
  let config;

  try {
    const { data } = await fetchRepoFile<ProjectTemplateRepoConfig>(
      HUBSPOT_PROJECT_COMPONENTS_GITHUB_PATH,
      'config.json',
      githubRef
    );
    config = data;
  } catch (err) {
    debugError(err);
  }

  if (config) {
    return config[PROJECT_COMPONENT_TYPES.COMPONENTS] || [];
  }
  return [];
}

export async function getProjectTemplateListFromRepo(
  templateSource: RepoPath,
  githubRef: string
): Promise<ProjectTemplate[]> {
  let config: ProjectTemplateRepoConfig;

  try {
    const { data } = await fetchRepoFile<ProjectTemplateRepoConfig>(
      templateSource,
      'config.json',
      githubRef
    );
    config = data;
  } catch (e) {
    debugError(e);
    logger.error(i18n(`${i18nKey}.errors.missingConfigFileTemplateSource`));
    return process.exit(EXIT_CODES.ERROR);
  }

  if (!config || !config[PROJECT_COMPONENT_TYPES.PROJECTS]) {
    logger.error(i18n(`${i18nKey}.errors.noProjectsInConfig`));
    return process.exit(EXIT_CODES.ERROR);
  }

  const templates = config[PROJECT_COMPONENT_TYPES.PROJECTS]!;

  const templatesContainAllProperties = templates.every(config =>
    PROJECT_TEMPLATE_PROPERTIES.every(p =>
      Object.prototype.hasOwnProperty.call(config, p)
    )
  );

  if (!templatesContainAllProperties) {
    logger.error(i18n(`${i18nKey}.errors.missingPropertiesInConfig`));
    return process.exit(EXIT_CODES.ERROR);
  }

  return templates;
}
