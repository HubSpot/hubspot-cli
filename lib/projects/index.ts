import fs from 'fs-extra';
import path from 'path';
import findup from 'findup-sync';
import { logger } from '@hubspot/local-dev-lib/logger';
import {
  cloneGithubRepo,
  fetchFileFromRepository,
} from '@hubspot/local-dev-lib/github';
import {
  createProject,
  fetchProject,
} from '@hubspot/local-dev-lib/api/projects';
import { isSpecifiedError } from '@hubspot/local-dev-lib/errors/index';
import { getCwd, getAbsoluteFilePath } from '@hubspot/local-dev-lib/path';
import { RepoPath } from '@hubspot/local-dev-lib/types/Github';
import { Project } from '@hubspot/local-dev-lib/types/Project';
import { HubSpotPromise } from '@hubspot/local-dev-lib/types/Http';

import {
  FEEDBACK_INTERVAL,
  DEFAULT_POLLING_DELAY,
  PROJECT_CONFIG_FILE,
  HUBSPOT_PROJECT_COMPONENTS_GITHUB_PATH,
  PROJECT_COMPONENT_TYPES,
} from '../constants';
import { promptUser } from '../prompts/promptUtils';
import { EXIT_CODES } from '../enums/exitCodes';
import { uiLine, uiAccountDescription, uiCommandReference } from '../ui';
import { i18n } from '../lang';
import SpinniesManager from '../ui/SpinniesManager';
import {
  ProjectTemplate,
  ProjectConfig,
  ProjectAddComponentData,
  ProjectTemplateRepoConfig,
  ComponentTemplate,
} from '../../types/Projects';
import { logError, ApiErrorContext } from '../errorHandlers/index';

const i18nKey = 'lib.projects';

export function writeProjectConfig(
  configPath: string,
  config: ProjectConfig
): boolean {
  try {
    fs.ensureFileSync(configPath);
    fs.writeFileSync(configPath, JSON.stringify(config, null, 2));
    logger.debug(`Wrote project config at ${configPath}`);
  } catch (e) {
    logger.debug(e);
    return false;
  }
  return true;
}

export function getIsInProject(dir?: string): boolean {
  const configPath = getProjectConfigPath(dir);
  return !!configPath;
}

function getProjectConfigPath(dir?: string): string | null {
  const projectDir = dir ? getAbsoluteFilePath(dir) : getCwd();

  const configPath = findup(PROJECT_CONFIG_FILE, {
    cwd: projectDir,
    nocase: true,
  });

  return configPath;
}

export async function getProjectConfig(dir?: string): Promise<{
  projectDir: string | null;
  projectConfig: ProjectConfig | null;
}> {
  const configPath = await getProjectConfigPath(dir);
  if (!configPath) {
    return { projectConfig: null, projectDir: null };
  }

  try {
    const config = fs.readFileSync(configPath);
    const projectConfig: ProjectConfig = JSON.parse(config.toString());
    return {
      projectDir: path.dirname(configPath),
      projectConfig,
    };
  } catch (e) {
    logger.error('Could not read from project config');
    return { projectConfig: null, projectDir: null };
  }
}

export async function createProjectConfig(
  projectPath: string,
  projectName: string,
  template: ProjectTemplate,
  templateSource: RepoPath,
  githubRef: string
): Promise<boolean> {
  const { projectConfig, projectDir } = await getProjectConfig(projectPath);

  if (projectConfig) {
    logger.warn(
      projectPath === projectDir
        ? 'A project already exists in that location.'
        : `Found an existing project definition in ${projectDir}.`
    );

    const { shouldContinue } = await promptUser<{ shouldContinue: boolean }>([
      {
        name: 'shouldContinue',
        message: () => {
          return projectPath === projectDir
            ? 'Do you want to overwrite the existing project definition with a new one?'
            : `Continue creating a new project in ${projectPath}?`;
        },
        type: 'confirm',
        default: false,
      },
    ]);

    if (!shouldContinue) {
      return false;
    }
  }

  const projectConfigPath = path.join(projectPath, PROJECT_CONFIG_FILE);

  logger.log(
    `Creating project config in ${
      projectPath ? projectPath : 'the current folder'
    }`
  );

  const hasCustomTemplateSource = Boolean(templateSource);

  await cloneGithubRepo(
    templateSource || HUBSPOT_PROJECT_COMPONENTS_GITHUB_PATH,
    projectPath,
    {
      sourceDir: template.path,
      tag: hasCustomTemplateSource ? undefined : githubRef,
    }
  );

  if (fs.existsSync(projectConfigPath)) {
    const _config: ProjectConfig = JSON.parse(
      fs.readFileSync(projectConfigPath).toString()
    );
    writeProjectConfig(projectConfigPath, {
      ..._config,
      name: projectName,
    });
  }

  if (template.name === 'no-template') {
    fs.ensureDirSync(path.join(projectPath, 'src'));
  }

  return true;
}

export function validateProjectConfig(
  projectConfig: ProjectConfig,
  projectDir: string
): void {
  if (!projectConfig) {
    logger.error(
      i18n(`${i18nKey}.validateProjectConfig.configNotFound`, {
        createCommand: uiCommandReference('hs project create'),
      })
    );
    return process.exit(EXIT_CODES.ERROR);
  }

  if (!projectConfig.name || !projectConfig.srcDir) {
    logger.error(i18n(`${i18nKey}.validateProjectConfig.configMissingFields`));
    return process.exit(EXIT_CODES.ERROR);
  }

  const resolvedPath = path.resolve(projectDir, projectConfig.srcDir);
  if (!resolvedPath.startsWith(projectDir)) {
    const projectConfigFile = path.relative(
      '.',
      path.join(projectDir, PROJECT_CONFIG_FILE)
    );
    logger.error(
      i18n(`${i18nKey}.validateProjectConfig.srcOutsideProjectDir`, {
        srcDir: projectConfig.srcDir,
        projectConfig: projectConfigFile,
      })
    );
    return process.exit(EXIT_CODES.ERROR);
  }

  if (!fs.existsSync(resolvedPath)) {
    logger.error(
      i18n(`${i18nKey}.validateProjectConfig.srcDirNotFound`, {
        srcDir: projectConfig.srcDir,
        projectDir: projectDir,
      })
    );

    return process.exit(EXIT_CODES.ERROR);
  }
}

async function pollFetchProject(
  accountId: number,
  projectName: string
): HubSpotPromise<Project> {
  // Temporary solution for gating slowness. Retry on 403 statusCode
  return new Promise((resolve, reject) => {
    let pollCount = 0;
    SpinniesManager.init();
    SpinniesManager.add('pollFetchProject', {
      text: i18n(`${i18nKey}.pollFetchProject.checkingProject`, {
        accountIdentifier: uiAccountDescription(accountId),
      }),
    });
    const pollInterval = setInterval(async () => {
      try {
        const response = await fetchProject(accountId, projectName);
        if (response && response.data) {
          SpinniesManager.remove('pollFetchProject');
          clearInterval(pollInterval);
          resolve(response);
        }
      } catch (err) {
        if (
          isSpecifiedError(err, {
            statusCode: 403,
            category: 'GATED',
            subCategory: 'BuildPipelineErrorType.PORTAL_GATED',
          }) &&
          pollCount < 15
        ) {
          pollCount += 1;
        } else {
          SpinniesManager.remove('pollFetchProject');
          clearInterval(pollInterval);
          reject(err);
        }
      }
    }, DEFAULT_POLLING_DELAY);
  });
}

export async function ensureProjectExists(
  accountId: number,
  projectName: string,
  {
    forceCreate = false,
    allowCreate = true,
    noLogs = false,
    withPolling = false,
    uploadCommand = false,
  } = {}
): Promise<{
  projectExists: boolean;
  project?: Project;
}> {
  const accountIdentifier = uiAccountDescription(accountId);
  try {
    const { data: project } = withPolling
      ? await pollFetchProject(accountId, projectName)
      : await fetchProject(accountId, projectName);
    return { projectExists: !!project, project };
  } catch (err) {
    if (isSpecifiedError(err, { statusCode: 404 })) {
      let shouldCreateProject = forceCreate;
      if (allowCreate && !shouldCreateProject) {
        const promptKey = uploadCommand ? 'createPromptUpload' : 'createPrompt';
        const promptResult = await promptUser<{ shouldCreateProject: boolean }>(
          [
            {
              name: 'shouldCreateProject',
              message: i18n(`${i18nKey}.ensureProjectExists.${promptKey}`, {
                projectName,
                accountIdentifier,
              }),
              type: 'confirm',
            },
          ]
        );
        shouldCreateProject = promptResult.shouldCreateProject;
      }

      if (shouldCreateProject) {
        try {
          const { data: project } = await createProject(accountId, projectName);
          logger.success(
            i18n(`${i18nKey}.ensureProjectExists.createSuccess`, {
              projectName,
              accountIdentifier,
            })
          );
          return { projectExists: true, project };
        } catch (err) {
          logError(err, new ApiErrorContext({ accountId }));
          return { projectExists: false };
        }
      } else {
        if (!noLogs) {
          logger.log(
            i18n(`${i18nKey}.ensureProjectExists.notFound`, {
              projectName,
              accountIdentifier,
            })
          );
        }
        return { projectExists: false };
      }
    }
    logError(err, new ApiErrorContext({ accountId }));
    process.exit(EXIT_CODES.ERROR);
  }
}

export function logFeedbackMessage(buildId: number): void {
  if (buildId > 0 && buildId % FEEDBACK_INTERVAL === 0) {
    uiLine();
    logger.log(i18n(`${i18nKey}.logFeedbackMessage.feedbackHeader`));
    uiLine();
    logger.log(i18n(`${i18nKey}.logFeedbackMessage.feedbackMessage`));
  }
}

export async function createProjectComponent(
  component: ProjectAddComponentData,
  name: string,
  projectComponentsVersion: string
): Promise<void> {
  const i18nKey = 'commands.project.subcommands.add';
  const componentName = name;

  const configInfo = await getProjectConfig();

  if (!configInfo.projectDir || !configInfo.projectConfig) {
    logger.error(i18n(`${i18nKey}.error.locationInProject`));
    process.exit(EXIT_CODES.ERROR);
  }

  const componentPath = path.join(
    configInfo.projectDir,
    configInfo.projectConfig.srcDir,
    component.insertPath,
    componentName
  );

  await cloneGithubRepo(HUBSPOT_PROJECT_COMPONENTS_GITHUB_PATH, componentPath, {
    sourceDir: component.path,
    tag: projectComponentsVersion,
  });
}

export async function getProjectComponentsByVersion(
  projectComponentsVersion: string
): Promise<ComponentTemplate[]> {
  const config = await fetchFileFromRepository<ProjectTemplateRepoConfig>(
    HUBSPOT_PROJECT_COMPONENTS_GITHUB_PATH,
    'config.json',
    projectComponentsVersion
  );

  return config[PROJECT_COMPONENT_TYPES.COMPONENTS] || [];
}
