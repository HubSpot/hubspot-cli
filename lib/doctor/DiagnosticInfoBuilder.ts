import { getProjectConfig } from '../projects/config.js';
import { fetchProject } from '@hubspot/local-dev-lib/api/projects';
import path from 'path';
import { pkg } from '../jsonLoader.js';
import { uiLogger } from '../ui/logger.js';
import { Environment } from '@hubspot/local-dev-lib/types/Config';
import {
  AccessToken,
  AccountType,
  AuthType,
} from '@hubspot/local-dev-lib/types/Accounts';
import { Project } from '@hubspot/local-dev-lib/types/Project';
import {
  getAccountId,
  getDefaultAccountOverrideFilePath,
  isConfigFlagEnabled,
} from '@hubspot/local-dev-lib/config';
import { getAccountConfig, getConfigPath } from '@hubspot/local-dev-lib/config';
import { getAccessToken } from '@hubspot/local-dev-lib/personalAccessKey';
import { walk } from '@hubspot/local-dev-lib/fs';
import util from 'util';
import { exec as execAsync } from 'node:child_process';
import process from 'process';
import {
  CMS_ASSETS_FILE,
  LEGACY_PRIVATE_APP_FILE,
  LEGACY_PUBLIC_APP_FILE,
  LEGACY_SERVERLESS_FILE,
  PROJECT_CONFIG_FILE,
  THEME_FILE,
} from '../constants.js';

export type ProjectConfig = Awaited<ReturnType<typeof getProjectConfig>>;

// This needs to be hardcoded since we are using it in the TS type
const hubspotCli = '@hubspot/cli';

interface FilesInfo {
  files: string[];
  configFiles: string[];
  packageFiles: string[];
  packageLockFiles: string[];
  envFiles: string[];
  jsonFiles: string[];
}

export interface DiagnosticInfo extends FilesInfo {
  path?: string;
  versions: { [hubspotCli]: string; node: string; npm: string | null };
  config: string | null;
  defaultAccountOverrideFile: string | null | undefined;
  configSettings: { [key: string]: unknown };
  project: {
    details?: Project;
    config?: ProjectConfig;
  };
  arch: typeof process.arch;
  platform: typeof process.platform;
  account: {
    name?: string;
    accountId?: number | null;
    scopeGroups?: AccessToken['scopeGroups'];
    enabledFeatures?: AccessToken['enabledFeatures'];
    accountType?: AccessToken['accountType'];
    authType?: AuthType;
  };
  diagnosis?: string;
  errorCount?: number;
  warningCount?: number;
}

const configFiles = [
  LEGACY_SERVERLESS_FILE,
  PROJECT_CONFIG_FILE,
  LEGACY_PRIVATE_APP_FILE,
  LEGACY_PUBLIC_APP_FILE,
  THEME_FILE,
  CMS_ASSETS_FILE,
];

export class DiagnosticInfoBuilder {
  accountId: number | null;
  readonly configSettings: { [key: string]: unknown };
  readonly env?: Environment;
  readonly authType?: AuthType;
  readonly accountType?: AccountType;
  readonly personalAccessKey?: string;
  private _projectConfig?: ProjectConfig;
  private accessToken?: AccessToken;
  private projectDetails?: Project;
  private files?: string[];
  readonly processInfo: NodeJS.Process;

  constructor(processInfo: NodeJS.Process) {
    this.accountId = getAccountId();
    const accountConfig = getAccountConfig(this.accountId!);
    this.configSettings = {
      httpUseLocalhost: isConfigFlagEnabled('httpUseLocalhost'),
    };
    this.env = accountConfig?.env;
    this.authType = accountConfig?.authType;
    this.accountType = accountConfig?.accountType;
    this.personalAccessKey = accountConfig?.personalAccessKey;
    this.processInfo = processInfo;
  }

  async generateDiagnosticInfo(): Promise<DiagnosticInfo> {
    this._projectConfig = await getProjectConfig();

    if (this._projectConfig?.projectConfig) {
      await this.fetchProjectDetails();
      await this.fetchAccessToken();
    }

    if (this._projectConfig?.projectDir) {
      await this.fetchProjectFilenames();
    }

    const {
      platform,
      arch,
      versions: { node },
      mainModule,
    } = this.processInfo;

    return {
      platform,
      arch,
      path: mainModule?.path,
      config: getConfigPath(),
      defaultAccountOverrideFile: getDefaultAccountOverrideFilePath(),
      configSettings: this.configSettings,
      versions: {
        [hubspotCli]: pkg.version,
        node,
        npm: await this.getNpmVersion(),
      },
      account: {
        accountId: this.accountId,
        accountType: this.accountType,
        authType: this.authType,
        name: this.accessToken?.hubName,
        scopeGroups: this.accessToken?.scopeGroups,
        enabledFeatures: this.accessToken?.enabledFeatures,
      },
      project: {
        config: this._projectConfig,
        details: this.projectDetails,
      },
      ...this.generateFilesArrays(),
      files: this.files || [],
    };
  }

  private async fetchProjectDetails(): Promise<void> {
    try {
      const { data } = await fetchProject(
        this.accountId!,
        // We check that config exists before running this function
        this._projectConfig!.projectConfig!.name
      );
      this.projectDetails = data;
    } catch (e) {
      uiLogger.debug(e);
    }
  }

  async fetchAccessToken(): Promise<AccessToken | undefined> {
    try {
      this.accessToken = await getAccessToken(
        this.personalAccessKey!,
        this.env,
        this.accountId!
      );
    } catch (e) {
      uiLogger.debug(e);
    }

    return this.accessToken;
  }

  private async fetchProjectFilenames(): Promise<void> {
    try {
      // We check that projectDir exists before running this function
      this.files = (await walk(this._projectConfig!.projectDir!))
        .filter(file => !path.dirname(file).includes('node_modules'))
        .map(filename =>
          path.relative(this._projectConfig!.projectDir!, filename)
        );
    } catch (e) {
      uiLogger.debug(e);
    }
  }

  private async getNpmVersion(): Promise<string | null> {
    const exec = util.promisify(execAsync);
    try {
      const { stdout } = await exec('npm --version');
      return stdout.toString().trim();
    } catch (e) {
      uiLogger.debug(e);
      return null;
    }
  }

  private generateFilesArrays(): FilesInfo {
    const output: FilesInfo = {
      files: this.files || [],
      configFiles: [],
      packageFiles: [],
      packageLockFiles: [],
      envFiles: [],
      jsonFiles: [],
    };

    if (!this.files) {
      return output;
    }

    return this.files.reduce((acc: FilesInfo, file) => {
      const { base } = path.parse(file);

      if (base === 'package.json') {
        acc.packageFiles.push(file);
      } else if (configFiles.includes(base)) {
        acc.configFiles.push(file);
        if (file.endsWith('.json')) {
          acc.jsonFiles.push(file);
        }
      } else if (base === 'package-lock.json') {
        acc.packageLockFiles.push(file);
      } else if (file.endsWith('.env')) {
        acc.envFiles.push(file);
      } else if (file.endsWith('.json')) {
        acc.jsonFiles.push(file);
      }

      return acc;
    }, output);
  }
}
