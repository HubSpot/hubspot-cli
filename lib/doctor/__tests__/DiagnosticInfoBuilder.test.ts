import util from 'util';

jest.mock('@hubspot/local-dev-lib/fs');
jest.mock('@hubspot/local-dev-lib/config');
jest.mock('@hubspot/local-dev-lib/personalAccessKey');
jest.mock('../../projects/config');
jest.mock('@hubspot/local-dev-lib/api/projects');
jest.mock('util', () => ({
  ...jest.requireActual('util'),
  promisify: jest.fn().mockReturnValue(jest.fn()),
}));

jest.mock('../../../package.json', () => ({
  version: '1.0.0',
}));

import { DiagnosticInfoBuilder, ProjectConfig } from '../DiagnosticInfoBuilder';
import {
  getAccountId as _getAccountId,
  getAccountConfig as _getAccountConfig,
  getConfigPath as _getConfigPath,
  getDefaultAccountOverrideFilePath as _getDefaultAccountOverrideFilePath,
  isConfigFlagEnabled as _isConfigFlagEnabled,
} from '@hubspot/local-dev-lib/config';
import { getAccessToken as _getAccessToken } from '@hubspot/local-dev-lib/personalAccessKey';
import { walk as _walk } from '@hubspot/local-dev-lib/fs';
import { AccessToken, CLIAccount } from '@hubspot/local-dev-lib/types/Accounts';
import { getProjectConfig as _getProjectConfig } from '../../projects/config';
import { fetchProject as _fetchProject } from '@hubspot/local-dev-lib/api/projects';
import { Project } from '@hubspot/local-dev-lib/types/Project';
import { HubSpotPromise } from '@hubspot/local-dev-lib/types/Http';

const walk = _walk as jest.MockedFunction<typeof _walk>;
const getAccessToken = _getAccessToken as jest.MockedFunction<
  typeof _getAccessToken
>;
const getAccountConfig = _getAccountConfig as jest.MockedFunction<
  typeof _getAccountConfig
>;
const getConfigPath = _getConfigPath as jest.MockedFunction<
  typeof _getConfigPath
>;
const getDefaultAccountOverrideFilePath =
  _getDefaultAccountOverrideFilePath as jest.MockedFunction<
    typeof _getDefaultAccountOverrideFilePath
  >;
const getAccountId = _getAccountId as jest.MockedFunction<typeof _getAccountId>;
const getProjectConfig = _getProjectConfig as jest.MockedFunction<
  typeof _getProjectConfig
>;
const isConfigFlagEnabled = _isConfigFlagEnabled as jest.MockedFunction<
  typeof _isConfigFlagEnabled
>;
const fetchProject = _fetchProject as jest.MockedFunction<typeof _fetchProject>;

const utilPromisify = util.promisify as jest.MockedFunction<
  typeof util.promisify
>;

describe('lib/doctor/DiagnosticInfo', () => {
  const accountId = 898989;
  const accountConfig: CLIAccount = {
    env: 'prod',
    authType: 'personalaccesskey',
    accountType: 'STANDARD',
    personalAccessKey: 'super-secret-key',
  };
  const nodeVersion = 'v18.17.0';
  const processInfo = {
    platform: 'darwin',
    arch: 'x64',
    versions: { node: nodeVersion },
    mainModule: { path: '/path/to/main/module' },
  } as NodeJS.Process;

  const projectDir = '/Users/test/project';
  const projectFiles = [
    `${projectDir}/.gitignore`,
    `${projectDir}/README.md`,
    // Config files
    `${projectDir}/hsproject.json`,
    `${projectDir}/src/app/app.json`,
    `${projectDir}/src/app/public-app.json`,
    // Serverless files
    `${projectDir}/src/app/app.functions/.env`,
    `${projectDir}/src/app/app.functions/function.js`,
    `${projectDir}/src/app/app.functions/serverless.json`,
    `${projectDir}/src/app/app.functions/package.json`,
    `${projectDir}/src/app/app.functions/package-lock.json`,
    // Extension files
    `${projectDir}/src/app/extensions/extension.js`,
    `${projectDir}/src/app/extensions/extension.json`,
    `${projectDir}/src/app/extensions/package.json`,
    `${projectDir}/src/app/extension/package-lock.json`,
    `${projectDir}/src/app/app.functions/node_modules/axios`,
  ];

  beforeEach(() => {
    getAccountId.mockReturnValue(accountId);
    getAccountConfig.mockReturnValue(accountConfig);
    walk.mockResolvedValue(projectFiles);
    isConfigFlagEnabled.mockReturnValue(false);
  });

  it('should initialize the required state on creation', () => {
    const builder = new DiagnosticInfoBuilder(processInfo);

    expect(getAccountId).toHaveBeenCalledTimes(1);
    expect(getAccountConfig).toHaveBeenCalledTimes(1);

    expect(builder.accountId).toEqual(accountId);

    expect(builder.env).toEqual(accountConfig.env);
    expect(builder.authType).toEqual(accountConfig.authType);
    expect(builder.accountType).toEqual(accountConfig.accountType);
    expect(builder.personalAccessKey).toEqual(accountConfig.personalAccessKey);
    expect(builder.processInfo).toEqual(processInfo);
  });

  describe('generateDiagnosticInfo', () => {
    let builder: DiagnosticInfoBuilder;
    let projectConfig: ProjectConfig;
    let projectDetails: Project;
    let accessToken: AccessToken;

    const npmVersion = 'v8.17.0';
    const configPath = '/path/to/config';
    const defaultAccountOverrideFile =
      'path/to/default/account/override/.hsaccount';

    beforeEach(() => {
      builder = new DiagnosticInfoBuilder(processInfo);

      projectConfig = {
        projectDir,
        projectConfig: {
          name: 'My project',
          srcDir: 'project-dir',
          platformVersion: 'test',
        },
      };

      projectDetails = {
        createdAt: 12345,
        deletedAt: 0,
        deployedBuildId: 1,
        id: 8989898,
        isLocked: false,
        name: projectConfig!.projectConfig!.name,
        portalId: accountId,
        updatedAt: 12345,
      };

      accessToken = {
        accessToken: 'super-secret-dont-put-this-in-a-unit-test',
        accountType: 'STANDARD',
        encodedOAuthRefreshToken: '',
        expiresAt: '',
        hubName: projectConfig!.projectConfig!.name,
        portalId: accountId,
        scopeGroups: [],
        enabledFeatures: {},
      };

      getProjectConfig.mockResolvedValue(projectConfig);
      fetchProject.mockResolvedValue({
        data: projectDetails,
      } as unknown as HubSpotPromise<Project>);
      getAccessToken.mockResolvedValue(accessToken);
      getConfigPath.mockReturnValue(configPath);
      getDefaultAccountOverrideFilePath.mockReturnValue(
        defaultAccountOverrideFile
      );
      utilPromisify.mockReturnValue(jest.fn().mockResolvedValue(npmVersion));
    });

    it('should gather the required data and generate the diagnostic', async () => {
      const diagnosticInfo = await builder.generateDiagnosticInfo();

      expect(getProjectConfig).toHaveBeenCalledTimes(1);

      expect(fetchProject).toHaveBeenCalledTimes(1);
      expect(fetchProject).toHaveBeenCalledWith(
        accountId,
        projectConfig!.projectConfig!.name
      );

      expect(getAccessToken).toHaveBeenCalledTimes(1);
      expect(getAccessToken).toHaveBeenCalledWith(
        accountConfig.personalAccessKey,
        accountConfig.env,
        accountId
      );

      // @ts-expect-error Accessing private field
      expect(builder._projectConfig).toEqual(projectConfig);

      expect(diagnosticInfo).toMatchSnapshot();
    });

    it('should handle errors when fetching project details', async () => {
      fetchProject.mockRejectedValue(
        new Error('Failed to fetch project details')
      );

      const diagnosticInfo = await builder.generateDiagnosticInfo();

      expect(fetchProject).toHaveBeenCalledTimes(1);
      expect(diagnosticInfo.project.details).toBeUndefined();
    });

    it('should handle errors when fetching access token', async () => {
      getAccessToken.mockRejectedValue(
        new Error('Failed to fetch access token')
      );

      const diagnosticInfo = await builder.generateDiagnosticInfo();

      expect(getAccessToken).toHaveBeenCalledTimes(1);
      expect(diagnosticInfo.account.name).toBeUndefined();
      expect(diagnosticInfo.account.scopeGroups).toBeUndefined();
      expect(diagnosticInfo.account.enabledFeatures).toBeUndefined();
    });

    it('should handle errors when fetching project filenames', async () => {
      walk.mockRejectedValue(new Error('Failed to walk project directory'));

      const diagnosticInfo = await builder.generateDiagnosticInfo();

      expect(walk).toHaveBeenCalledTimes(1);
      expect(diagnosticInfo.files).toEqual([]);
    });
  });
});
