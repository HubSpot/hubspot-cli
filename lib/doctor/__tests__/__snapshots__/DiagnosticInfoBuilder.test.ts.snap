// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/doctor/DiagnosticInfo generateDiagnosticInfo should gather the required data and generate the diagnostic 1`] = `
{
  "account": {
    "accountId": 898989,
    "accountType": "STANDARD",
    "authType": "personalaccesskey",
    "enabledFeatures": {},
    "name": "My project",
    "scopeGroups": [],
  },
  "arch": "x64",
  "config": "/path/to/config",
  "configFiles": [
    "hsproject.json",
    "src/app/app.json",
    "src/app/public-app.json",
    "src/app/app.functions/serverless.json",
  ],
  "defaultAccountOverrideFile": "path/to/default/account/override/.hsaccount",
  "envFiles": [
    "src/app/app.functions/.env",
  ],
  "files": [
    ".gitignore",
    "README.md",
    "hsproject.json",
    "src/app/app.json",
    "src/app/public-app.json",
    "src/app/app.functions/.env",
    "src/app/app.functions/function.js",
    "src/app/app.functions/serverless.json",
    "src/app/app.functions/package.json",
    "src/app/app.functions/package-lock.json",
    "src/app/extensions/extension.js",
    "src/app/extensions/extension.json",
    "src/app/extensions/package.json",
    "src/app/extension/package-lock.json",
  ],
  "jsonFiles": [
    "hsproject.json",
    "src/app/app.json",
    "src/app/public-app.json",
    "src/app/app.functions/serverless.json",
    "src/app/extensions/extension.json",
  ],
  "packageFiles": [
    "src/app/app.functions/package.json",
    "src/app/extensions/package.json",
  ],
  "packageLockFiles": [
    "src/app/app.functions/package-lock.json",
    "src/app/extension/package-lock.json",
  ],
  "path": "/path/to/main/module",
  "platform": "darwin",
  "project": {
    "config": {
      "projectConfig": {
        "name": "My project",
        "platformVersion": "test",
        "srcDir": "project-dir",
      },
      "projectDir": "/Users/test/project",
    },
    "details": {
      "createdAt": 12345,
      "deletedAt": 0,
      "deployedBuildId": 1,
      "id": 8989898,
      "isLocked": false,
      "name": "My project",
      "portalId": 898989,
      "updatedAt": 12345,
    },
  },
  "versions": {
    "@hubspot/cli": "1.0.0",
    "node": "v18.17.0",
    "npm": null,
  },
}
`;
